New spends forward pass

wd60622 opened the pull request:
Current code addressing #451. Have a look @carlosagostini 

I've split it into two methods: 
- generation the data for the plot
- plotting of the data

I'd like to hear feedback on the arguments for the generation. Specifically, the `spends` and the `one_time` 
Currently, it takes a np.ndarray that would be a raw spend for each channel (error if it doesn't match) and will ultimately transform it to a (max_lag, n_channels) ndarray with values depending on the `one_time` argument. 
One time: 
- True -> spends are the first and all rest are zeros in the (max_lag, n_channels)
- False -> spends for each channel are for each of the 
This locks into a bit of assumption of how this would be used. Maybe, the spends could be (n_channels, ) or (max_lags, n_channels) to reduce the need of `one_time` and simplify the method.
Would like to hear how you might use it. 

The method returns the untransformed spends although named `_contributions`. The inverse transformation only happen in the plotting method, but it maybe good to have the inverse transformations happening that first method.
I think this is a good case to remove the target transformation from the sklearn pipeline so it can be part of the forward pass of contributions @ricardoV94

EDIT: The inverse transformation is done in the new_spend_contributions method as well

<!-- readthedocs-preview pymc-marketing start -->
----
:books: Documentation preview :books:: https://pymc-marketing--456.org.readthedocs.build/en/456/

<!-- readthedocs-preview pymc-marketing end -->

wd60622 commented:
Got the prior functionality working

<img width="1003" alt="prior-and-posterior-support" src="https://github.com/pymc-labs/pymc-marketing/assets/57733339/811c2ec5-735e-4336-9e18-5d59e58de999">

The inverse_transform could be used if the model was fit. Might be good to toggle inverse_transform even for the posterior. Any thoughts here?

wd60622 commented:
Will add the transformation back to original space and continue with this PR after #482

wd60622 commented:
Think that the API is finalized 
Will add to the docstrings and will add some tests tomorrow and work through any rebase issues

twiecki commented:
What's the status of this? Any blockers to moving this forward?

wd60622 commented:
> What's the status of this? Any blockers to moving this forward?

Have the main functionality built out
Next steps

- add method docstring for plotting
- round out test for non-plotting method
- review & feedback

wd60622 commented:
I'd appreciate any feedback on improving the `new_spend_contributions` function
I brought in `pytest-mpl` in order to test `plot_new_spend_contributions` but I think if `new_spend_contributions` is tested enough, it might be fine at the moment

Would love to hear any thoughts and feedback

juanitorduz reviewed (changes requested):
This looks very nice! I added some minor suggestions on the docstrings to clarify this better for the end-user. WDYT?

wd60622 commented on pymc_marketing/mmm/delayed_saturated_mmm.py:
```diff
@@ -850,6 +849,208 @@ def plot_channel_contributions_grid(
+            "channel": self.channel_columns,
+        }
+        with pm.Model(coords=coords):
+            alpha = pm.Uniform("alpha", lower=0, upper=1, dims=("channel",))
+            lam = pm.HalfFlat("lam", dims=("channel",))
+            beta_channel = pm.HalfFlat("beta_channel", dims=("channel",))
```
The are arbitrary and could all be Flat distributions. Only the shape is required to match in order to work with the `sample_posterior_predictive` method

wd60622 replied:
Might be good to get the dims from the config though...

wd60622 commented on pymc_marketing/mmm/delayed_saturated_mmm.py:
```diff
@@ -850,6 +849,208 @@ def plot_channel_contributions_grid(
+        )
+
+        contributions_groupby = contributions.to_series().groupby(
+            level=["weeks_since_spend", "channel"]
```
"time units since spend"? 😆

wd60622 commented on pymc_marketing/mmm/delayed_saturated_mmm.py:
```diff
@@ -850,6 +849,208 @@ def plot_channel_contributions_grid(
+            self.adstock_max_lag, -1
+        )
+
+        spend = (
+            np.vstack(
+                [spend, np.zeros((self.adstock_max_lag, len(self.channel_columns)))]
+            )
+            if one_time
+            else np.ones((self.adstock_max_lag + 1, len(self.channel_columns))) * spend
+        )
```
Good call. Will do!

wd60622 commented:
I've addressed all the feedback, @juanitorduz
Thanks!

I've separated out the "new_spend" creation into a function and added some more to the docstrings so hopefully that brings in some clarity. 

Let me know what you think!

juanitorduz commented:
@wd60622 looks great! There is one test failing though. If you fix it we can merge this one  :)

wd60622 commented:
Have the test passing locally but not in the action 😢  

Running the action with 3.11 with the latest pymc to see if the pymc version has an impact. The test depends on the sample_prior_predictive method. Not sure if that has changed

Might just mock the idata for all of these tests with the plots and hope giving a seed will keep the results the sample

wd60622 commented:
Trying to mock the prior in the `model.idata` 
Hope that that works. If not, I will pull out the plotting into a function and tests on some fixed data. 

If all else fails, these are tests against the plotting images and can maybe be omitted. Thoughts? @juanitorduz @ricardoV94

juanitorduz commented:
@wd60622 I am getting the same error as in https://github.com/matplotlib/pytest-mpl/issues/69 . Can you please add `markers =["mpl_image_compare"]` to pyproject.toml as in

```
[tool.pytest.ini_options]
addopts = [
    "-v",
    "--strict-markers",
    "--strict-config",
    "--cov=pymc_marketing",
    "--cov-report=term-missing",
    "--color=yes",
]
markers =["mpl_image_compare"]
filterwarnings = ["ignore::DeprecationWarning:bokeh.core.property.primitive:37"]
testpaths = "tests"
```

wd60622 commented:
Failed again even after using fresh environment 😢 
Would you be able to check one more time @juanitorduz

juanitorduz commented:
> Failed again even after using fresh environment 😢 Would you be able to check one more time @juanitorduz

I suggest we make sure the plot runs (as other plots tests) and create an issue to add this matplotlib testing framework. I would not delay this feature because of this :)

wd60622 commented:
> > Failed again even after using fresh environment 😢 Would you be able to check one more time @juanitorduz
> 
> I suggest we make sure the plot runs (as other plots tests) and create an issue to add this matplotlib testing framework. I would not delay this feature because of this :)

Sounds great!

juanitorduz commented:
Thanks @wd60622 !

wd60622 commented:
> Thanks @wd60622 !

Thank you for the review!
